### ISSUE TITLE
mysterious connection latency when building deb package of PR #5720

### ISSUE THREAD

**Issue by mbroadst** [rethinkdb/rethinkdb#5893]
Hey guys, I'm running out of ideas for how to fix this issue so I figured I'd post the question up as an issue here. Here's the rundown:

I'm building local debian packages of two branches of rethinkdb: v2.3.4 (vanilla) and a branch of v2.3.4 with a single squashed commit on it comprising all the changes in #5720. Both debian files are build as follows:
```
$ ./configure --allow-fetch --fetch protobuf --fetch v8 --fetch re2 --fetch z --fetch crypto --fetch ssl --fetch boost_system --fetch boost --fetch jemalloc --fetch curl

$ UBUNTU_RELEASE=trusty SIGN_PACKAGE=0 make build-deb
```

The test system is a single ubuntu 14.04 x86_64 system, not joined to any cluster. That last bit is incredibly important because if you take a look at #5720 again, you'll note that it only touches code that related to joining a cluster (either directly, or through the `auto_reconnector_t`). 

I run two test cases against the built debian packages with the following code:
```
const r = require('rethinkdbdash')();
setInterval(() => {
  r.db('rethinkdb').table('table_config')
    .then(data => console.log('worked: ', data.length))
    .catch(err => console.log('[interval]: ', err));
}, 3000);

// log uncaught exceptions and bail
function globalErrorHandler(err) {
  console.log('[global]: ', err);
  process.exit(1);
}

process.on('uncaughtException', globalErrorHandler);
process.on('unhandledRejection', globalErrorHandler);
```

This is using `rethinkdbdash` version `2.3.12`. (**NOTE:** Internally in its default configuration `rethinkdbdash` creates 50 connections in a pool to rethinkdb).

I'm finding some incredibly strange results, given the previous information:
* When run against v2.3.4, all is as expected, every 3s I get a message `worked: <some number>`
* When run against v2.3.4+PR changes, however, I get a few `worked: <some number>` messages before ~45 of these errors: `Fail to create a new connection for the connection pool. Error:{"msg":"Failed to connect to localhost:28015 in less than 20s","message":"Failed to connect to localhost:28015 in less than 20s.","isOperational":true}`
* MOST strange: when I manually build the PR code (read: _not_ the debian package), strip the executable and copy it over to the machine that has the PR debian package _or_ the vanilla package installed, it works perfectly as expected (including the added functionality of the PR)

It smells like the problem is in the debian packaging, but I've almost completely lost perspective on this issue having tried to figure out what's going on over the past few days. I've conducted this experiment on three different systems at this point, and can reproduce it on every one. 

I really hope one of you guys has an idea what's going on, because I'm pretty stumped.

**Comment by mlucy** [comment]
@AtnNn -- any idea about what part of the debian packaging process might cause this behavior?

**Comment by AtnNn** [comment]
@mlucy The debian package might get passed different `--fetch` options or be built with a different version string. It might also get stripped differently. The dpkg build process might also be using a different toolset or default compiler options.

It's possible there are interfering artifacts from previous builds. Perhaps running `make clean` or even `git clean -xd` before building might help.

@mbroadst What is the exact commit? After the connections start failing, is the server still running? Is it under high load or is it hung on IO? What does its log report?

**Comment by mbroadst** [comment]
@AtnNn yes I'm very careful to `git clean -dfx` before each build, and I've modified my testing process to configure with `--fetch all` as per your suggestion on slack. I agree with you assessment that the PPA might be building with different options, but I'm not sure it explains the fact that the vanilla `v2.3.4` builds and works fine (both remotely and locally built deb packages), as well as the fact that the `sid` (my changes from the PR) deb when built locally exhibits the same behavior as when built remotely on the PPA. 

There are three particular commits, you can see them on this branch: https://github.com/mbroadst/rethinkdb/tree/stop-reconnect-unknown-sid-on-v2.3.4. That code is branched off the `v2.3.4` tag, with the three commits from the PR on top of it. There are no logs of interest whatsoever when run, and I can't really see much of a cpu deviation when doing quick monitor with `htop`. It does feel like something is screwing with coro's internally, but I'm not quite sure how to prove that atm.

I would be incredibly interested to see if you guys get the same results as I do building a deb at your office, would that be possible?

**Comment by danielmewes** [comment]
Very mysterious. I definitely can't see anything in your commits that should affect normal (non-clustered) query behavior.

Just to make sure, could you try the experiment with e.g. only the first of your three commits applied?
The only other thing I can think of is that having additional commits after the last tag (v2.3.4) will change how the version number is generated. You can take a look at the scripts/gen-version.sh script, and maybe (for testing) make it output `2.3.4` instead of what it usually does. I have no clue how that would have any effect on this test though. Maybe it affects rethinkdbdash somehow.

**Comment by mbroadst** [comment]
@danielmewes I can certainly check it out. Would it be possible to try to reproduce it on your side of the equation as well? I did my tests on three separate machines with no success, and I sort of stopped working on it since then because it was slowly driving me mad 😄 

I do modify the `gen-version.sh` already, and simply add a `-sid` to the end of the version, but like you I can't think that would have anything to do with the runtime. I'm also pretty deeply familiar with the internals of rethinkdbdash and can't think of where that might affect anything on that side. It does seem more likely to me that it has something to do with stripping the binary, or rather something compiler related.
